### ISSUE TITLE
Advice for retraining on multiple samples

### ISSUE THREAD

**Issue by mpinese** [google/deepvariant#312]
Hello, after some quite impressive results applying HiSeq-trained DeepVariant on MGISEQ-2000 data, I've been working on achieving even better performance by retraining DeepVariant specifically for the MGISEQ-2000.

To do this I've been broadly following the sketch at https://github.com/google/deepvariant/blob/r0.10/docs/deepvariant-training-case-study.md.  I now have unshuffled tfrecords for six reference samples, and have some questions about next steps.

Because it might be relevant to the questions, my data are structured as follows:
* 6 samples (4xHG001/NA12878, 2xHG005/NA24631), each with:
  * 25 tfrecord shards (00000-00024) of chr1, for tuning (perhaps over-optimistically) 
  * 247 tfrecord shards (00000-00246) of chr2-19, no downsampling, for training
  * 247 tfrecord shards (00000-00246) of chr2-19, 50% downsampling, for training
  * 17 tfrecord shards (00000-00016) of chr20-22, for validation.

My questions are:
1. Is it necessary to shuffle the training data?  I ask as it's proving to be a bit laborious to set up, and so I'm hoping that I can get around it. Given I have so many shards, if I just shuffle the order of the chr2-19 shards when I supply them to the training loop, will this be almost as good as shuffling the whole dataset?
2. Is it necessary to shuffle the validation data?  The tutorial does this, but I'm not sure why.
3. How can I supply multiple datasets to the training loop (here effectively 12 datasets: 6 samples x 2 downsampling settings)?  In the tutorial, `model_train` is supplied a wildcard path of `validation_set.with_label.shuffled-?????-of-?????.tfrecord.gz`, which seems like it would only work for a single sample, and I'm not sure how this will work with multiple samples.
4. Have there been any changes to the code base to better support warmstarting, or is the advice at https://github.com/google/deepvariant/issues/185 still the best approach to fine-tuning the model?

DeepVariant is a fantastic tool and I'm very much looking forward to seeing what it can do with these data, so many thanks in advance.

**Comment by akolesnikov** [comment]
Hi Mark,

1. Shuffling is done so that each training batch is more representative of the entire data set. It helps the training to converge faster and avoid overfitting. If you shuffle the way you described it won't help. Although you may train without shuffling, it just won't work as well. Another problem is that without shuffling step you don't have all of your training data in one sharded file (see answer (3))
2. Shuffling is needed for tune and train data, and not needed for validation data.
3. input_pattern_list of shuffle script takes a list of files. This way you shuffle all of your samples into one sharded file. 
4. Will comment on this in another post.

**Comment by mpinese** [comment]
Thanks so much, that makes the path forwards a lot clearer.  Reading between the lines, is (3) the reason we are advised to shuffle the validation data as well, because it combines all the chr20-22 examples into one dataset?

Very much looking forward to reading your comments on warmstarting.

**Comment by pichuan** [comment]
Hi @mpinese ,
I'm commenting here for your question 4.

Internally for warmstarting, we use the same flag (`--start_from_checkpoint`) in https://github.com/google/deepvariant/blob/r0.10/docs/deepvariant-training-case-study.md  to train our WES and PacBio release models.

You refer to: https://github.com/google/deepvariant/issues/185 . The advice in my comment https://github.com/google/deepvariant/issues/185#issuecomment-494919509 has a code change that you could try out, only if you want to experiment with different warmstarting logic on your own. 

What we currently use is what you can find in our r0.10 codebase:
https://github.com/google/deepvariant/blob/4b937f03a1336d1dc6fd4c0eef727e1f83d2152a/deepvariant/modeling.py#L560


When warmstarting and training with a small(er) amount of data, it's always possible that the curve might look a bit weird at the beginning. Here is a rule of thumb I use: a stable training setup should be mostly reproducible. Meaning, if you run the same training multiple times, the curves should eventually converge to about the same place, and shouldn't behave drastically different. They won't look the exactly same because of randomness in training process. But if half of the runs don't converge, or behave very differently from the other half, then something needs to be improved.

You might also have a question on whether you need to warmstart. That is an empirical question. Here is an example from my experience:
For our PacBio training, at this point I actually feel like we have enough data to not have to warmstart from the WGS model. But we're still warmstarting (at least for now) because I find that it converges faster and the resulting accuracy is about the same. We make these decisions based on empirical evidence and our intuition on ML and the data. These decisions can also evolve over time.

( Btw, one thing that we might not have documented - if you want to try with *not* warmstarting from anything, and want to just randomly init, you can set `--start_from_checkpoint=""`. )


I think your setup above looks reasonable to me. If you want to share more later (like what your training looks like, how long it takes to converge, whether your new model works better on your data, etc), feel free to add more in this issue. I will close it for now.

**Comment by pichuan** [comment]
> Thanks so much, that makes the path forwards a lot clearer. Reading between the lines, is (3) the reason we are advised to shuffle the validation data as well, because it combines all the chr20-22 examples into one dataset?


And, I forgot to reply this part - for validation set, running the shuffling step is mainly for combing the examples into one dataset (and create a text file that describe it). Another reason is - if you specify `max_examples` in model_eval, it just might be safer to have the examples already pre-shuffled:
https://github.com/google/deepvariant/blob/4b937f03a1336d1dc6fd4c0eef727e1f83d2152a/deepvariant/model_eval.py#L118

---

One more thing to clarify:
Unfortunately I might be using terminology that are a bit confusing in the doc:

In https://github.com/google/deepvariant/blob/r0.10/docs/deepvariant-training-case-study.md , 
our **training** set are the labeled examples that our classifier actually learns from. 
This is the same as defined in: https://developers.google.com/machine-learning/glossary/#training_set

**Validation** set  is the labeled examples that our classifier never directly learns from, but is used to pick a best model checkpoint. 
This is the same as defined in: https://developers.google.com/machine-learning/glossary/#validation_set
When I wrote our training case study, I tried to not use another term "**tuning** set", which is the same thing as "validation set". But now I read it again, I think I did accidentally use the term "tuning set" at least once. I'll update that in a future release.
Just to clarify again: When I use the term "tuning set", I'm referring to the same thing as validation set.

There is also **test** set, defined here: https://developers.google.com/machine-learning/glossary/#test_set
When doing machine learning, the best practice is to leave the **test** set alone as much as possible, and NOT to make any decisions (on model picking, paramater tuning) on it at all. Which is why we had a validation set held out in the first place.
